src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,313 @@
+#include <Common/typeid_cast.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Interpreters/ArithmeticOperationsInAgrFuncOptimize.h>
+#include <IO/WriteHelpers.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int UNEXPECTED_AST_STRUCTURE;
+}
+
+static constexpr const char * min = "min";
+static constexpr const char * max = "max";
+static constexpr const char * mul = "multiply";
+static constexpr const char * plus = "plus";
+static constexpr const char * sum = "sum";
+
+bool isConstantField(const Field & field)
+{
+    return field.getType() == Field::Types::Int64 ||
+           field.getType() == Field::Types::UInt64 ||
+           field.getType() == Field::Types::Int128 ||
+           field.getType() == Field::Types::UInt128;
+}
+
+bool onlyConstsInside(ASTFunction * func_node)
```

Comment:
Here and further. If pointer could be const if should be const: `const ASTFunction *`